File: api/routers/security_monitor.py
"""
Security Monitor Router
Monitoramento de segurança em tempo real
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from datetime import datetime, timedelta
from typing import List, Dict, Any
import json
import os
from pathlib import Path

from database import get_db
from auth import get_current_user

router = APIRouter()

# Caminho para logs e arquivos de segurança
SECURITY_DIR = Path(__file__).parent.parent.parent / "security"
CHECKSUMS_FILE = Path(__file__).parent.parent.parent / "checksums.json"
SCAN_RESULTS_FILE = SECURITY_DIR / "scan_results.json"

@router.get("/security/status")
async def get_security_status(current_user: dict = Depends(get_current_user)):
    """
    Retorna status geral de segurança
    """
    
    # Verificar se WAF está realmente ativo
    waf_active = False
    try:
        # Verificar se WAF está habilitado no main.py
        import sys
        from pathlib import Path
        main_path = Path(__file__).parent.parent / "main.py"
        with open(main_path, 'r', encoding='utf-8') as f:
            main_content = f.read()
            # Verificar se a linha do WAF não está comentada
            if 'app.add_middleware(WAFMiddleware)' in main_content and \
               not '# app.add_middleware(WAFMiddleware)' in main_content:
                waf_active = True
    except:
        pass
    
    status = {
        "timestamp": datetime.now().isoformat(),
        "waf": {
            "status": "active" if waf_active else "disabled",
            "protections": [
                "SQL Injection",
                "XSS",
                "Rate Limiting",
                "IP Blacklist"
            ]
        },
        "integrity": {
            "status": "unknown",
            "last_check": None,
            "files_monitored": 0
        },
        "vulnerabilities": {
            "status": "unknown",
            "last_scan": None,
            "issues_found": 0
        },
        "compliance": {
            "lgpd": "compliant",
            "iso27001": "compliant",
            "owasp": "compliant"
        }
    }
    
    # Verificar checksums
    if CHECKSUMS_FILE.exists():
        try:
            with open(CHECKSUMS_FILE, 'r') as f:
                checksums_data = json.load(f)
                status["integrity"]["status"] = "monitored"
                status["integrity"]["last_check"] = checksums_data.get("generated_at")
                status["integrity"]["files_monitored"] = checksums_data.get("total_files", 0)
        except:
            pass
    
    # Verificar scan results
    if SCAN_RESULTS_FILE.exists():
        try:
            with open(SCAN_RESULTS_FILE, 'r') as f:
                scan_data = json.load(f)
                status["vulnerabilities"]["last_scan"] = scan_data.get("scan_date")
                
                # Contar vulnerabilidades
                issues = 0
                for key in ["python", "nodejs", "docker"]:
                    if scan_data.get(key, {}).get("status") == "vulnerabilities_found":
                        vulns = scan_data[key].get("vulnerabilities", [])
                        if isinstance(vulns, list):
                            issues += len(vulns)
                        elif isinstance(vulns, dict):
                            issues += sum(vulns.values())
                
                status["vulnerabilities"]["issues_found"] = issues
                status["vulnerabilities"]["status"] = "clean" if issues == 0 else "issues_found"
        except:
            pass
    
    return status


@router.get("/security/waf/stats")
async def get_waf_stats(current_user: dict = Depends(get_current_user)):
    """
    Retorna estatísticas do WAF
    """
    
    # TODO: Implementar coleta de métricas do WAF
    # Por enquanto, retornar dados simulados
    
    return {
        "timestamp": datetime.now().isoformat(),
        "requests_blocked": 0,
        "sql_injection_attempts": 0,
        "xss_attempts": 0,
        "rate_limit_violations": 0,
        "blacklisted_ips": 0,
        "total_requests": 0,
        "uptime": "active"
    }


@router.get("/security/integrity/status")
async def get_integrity_status(current_user: dict = Depends(get_current_user)):
    """
    Retorna status da verificação de integridade
    """
    
    if not CHECKSUMS_FILE.exists():
        return {
            "status": "not_configured",
            "message": "Checksums not generated. Run: python security/integrity_check.py generate"
        }
    
    try:
        with open(CHECKSUMS_FILE, 'r') as f:
            data = json.load(f)
        
        return {
            "status": "configured",
            "generated_at": data.get("generated_at"),
            "total_files": data.get("total_files", 0),
            "base_directory": data.get("base_directory"),
            "last_verification": None  # TODO: Armazenar última verificação
        }
    except Exception as e:
        return {
            "status": "error",
            "message": str(e)
        }


@router.post("/security/integrity/generate")
async def generate_checksums(current_user: dict = Depends(get_current_user)):
    """Gera checksums de integridade executando o script diretamente"""
    import subprocess
    import sys
    try:
        script_path = Path(__file__).parent.parent.parent / "security" / "integrity_check.py"
        result = subprocess.run(
            [sys.executable, str(script_path), "generate"],
            capture_output=True, text=True, timeout=60,
            cwd=str(Path(__file__).parent.parent.parent)
        )
        if result.returncode == 0:
            return {"status": "success", "message": "Checksums gerados com sucesso", "output": result.stdout[-500:]}
        else:
            return {"status": "error", "message": result.stderr[-500:] or result.stdout[-500:]}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/security/vulnerabilities/scan")
async def run_vulnerability_scan(current_user: dict = Depends(get_current_user)):
    """Executa scan de vulnerabilidades"""
    import subprocess
    import sys
    try:
        script_path = Path(__file__).parent.parent.parent / "security" / "scan_dependencies.py"
        result = subprocess.run(
            [sys.executable, str(script_path)],
            capture_output=True, text=True, timeout=120,
            cwd=str(Path(__file__).parent.parent.parent)
        )
        if result.returncode == 0:
            return {"status": "success", "message": "Scan concluído", "output": result.stdout[-500:]}
        else:
            return {"status": "error", "message": result.stderr[-500:] or result.stdout[-500:]}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/security/integrity/verify")
async def verify_integrity(current_user: dict = Depends(get_current_user)):
    """
    Executa verificação de integridade
    """
    
    if not CHECKSUMS_FILE.exists():
        raise HTTPException(
            status_code=400,
            detail="Checksums not generated. Run: python security/integrity_check.py generate"
        )
    
    # TODO: Executar verificação em background
    # Por enquanto, retornar instrução
    
    return {
        "status": "pending",
        "message": "Run: python security/integrity_check.py verify",
        "note": "Background verification will be implemented in future version"
    }


@router.get("/security/vulnerabilities/status")
async def get_vulnerabilities_status(current_user: dict = Depends(get_current_user)):
    """
    Retorna status do scan de vulnerabilidades
    """
    
    if not SCAN_RESULTS_FILE.exists():
        return {
            "status": "not_scanned",
            "message": "No scan results found. Run: python security/scan_dependencies.py"
        }
    
    try:
        with open(SCAN_RESULTS_FILE, 'r') as f:
            data = json.load(f)
        
        return {
            "status": "scanned",
            "scan_date": data.get("scan_date"),
            "python": data.get("python", {}),
            "nodejs": data.get("nodejs", {}),
            "docker": data.get("docker", {})
        }
    except Exception as e:
        return {
            "status": "error",
            "message": str(e)
        }


@router.post("/security/scan")
async def run_security_scan(current_user: dict = Depends(get_current_user)):
    """
    Executa scan de segurança
    """
    
    # TODO: Executar scan em background
    # Por enquanto, retornar instrução
    
    return {
        "status": "pending",
        "message": "Run: python security/scan_dependencies.py",
        "note": "Background scanning will be implemented in future version"
    }


@router.get("/security/logs")
async def get_security_logs(
    limit: int = 100,
    current_user: dict = Depends(get_current_user)
):
    """
    Retorna logs de segurança recentes
    """
    
    # TODO: Implementar leitura de logs do WAF
    # Por enquanto, retornar lista vazia
    
    return {
        "timestamp": datetime.now().isoformat(),
        "logs": [],
        "total": 0,
        "note": "Log collection will be implemented in future version"
    }


@router.get("/security/recommendations")
async def get_security_recommendations(current_user: dict = Depends(get_current_user)):
    """
    Retorna recomendações de segurança
    """
    
    recommendations = []
    
    # Verificar checksums
    if not CHECKSUMS_FILE.exists():
        recommendations.append({
            "priority": "high",
            "category": "integrity",
            "title": "Generate File Checksums",
            "description": "File integrity monitoring is not configured",
            "action": "Run: python security/integrity_check.py generate"
        })
    
    # Verificar scan
    if not SCAN_RESULTS_FILE.exists():
        recommendations.append({
            "priority": "medium",
            "category": "vulnerabilities",
            "title": "Run Security Scan",
            "description": "No vulnerability scan has been performed",
            "action": "Run: python security/scan_dependencies.py"
        })
    else:
        # Verificar se scan é recente (últimas 7 dias)
        try:
            with open(SCAN_RESULTS_FILE, 'r') as f:
                data = json.load(f)
                scan_date = datetime.fromisoformat(data.get("scan_date"))
                if datetime.now() - scan_date > timedelta(days=7):
                    recommendations.append({
                        "priority": "medium",
                        "category": "vulnerabilities",
                        "title": "Update Security Scan",
                        "description": f"Last scan was {(datetime.now() - scan_date).days} days ago",
                        "action": "Run: python security/scan_dependencies.py"
                    })
        except:
            pass
    
    # Recomendações gerais
    recommendations.append({
        "priority": "low",
        "category": "maintenance",
        "title": "Regular Security Audits",
        "description": "Perform security audits monthly",
        "action": "Schedule regular security reviews"
    })
    
    return {
        "timestamp": datetime.now().isoformat(),
        "total": len(recommendations),
        "recommendations": recommendations
    }

File: frontend/src/components/SecurityMonitor.js
import React, { useState, useEffect } from 'react';
import api from '../services/api';
import './SecurityMonitor.css';

const SecurityMonitor = () => {
  const [status, setStatus] = useState(null);
  const [wafStats, setWafStats] = useState(null);
  const [integrityStatus, setIntegrityStatus] = useState(null);
  const [vulnerabilities, setVulnerabilities] = useState(null);
  const [recommendations, setRecommendations] = useState([]);
  const [loading, setLoading] = useState(true);
  const [activeTab, setActiveTab] = useState('overview');

  useEffect(() => {
    loadSecurityData();
    const interval = setInterval(loadSecurityData, 30000); // Atualizar a cada 30s
    return () => clearInterval(interval);
  }, []);

  const loadSecurityData = async () => {
    try {
      // Carregar status geral
      const statusRes = await api.get('/security/status');
      setStatus(statusRes.data);

      // Carregar estatísticas do WAF
      const wafRes = await api.get('/security/waf/stats');
      setWafStats(wafRes.data);

      // Carregar status de integridade
      const integrityRes = await api.get('/security/integrity/status');
      setIntegrityStatus(integrityRes.data);

      // Carregar vulnerabilidades
      const vulnRes = await api.get('/security/vulnerabilities/status');
      setVulnerabilities(vulnRes.data);

      // Carregar recomendações
      const recRes = await api.get('/security/recommendations');
      setRecommendations(recRes.data.recommendations || []);

      setLoading(false);
    } catch (error) {
      console.error('Error loading security data:', error);
      setLoading(false);
    }
  };

  const getStatusIcon = (status) => {
    switch (status) {
      case 'active':
      case 'compliant':
      case 'clean':
      case 'configured':
        return '✅';
      case 'issues_found':
      case 'vulnerabilities_found':
        return '⚠️';
      case 'unknown':
      case 'not_configured':
      case 'not_scanned':
        return '❓';
      default:
        return '⚠️';
    }
  };

  const getPriorityColor = (priority) => {
    switch (priority) {
      case 'high':
        return '#dc3545';
      case 'medium':
        return '#ffc107';
      case 'low':
        return '#17a2b8';
      default:
        return '#6c757d';
    }
  };

  if (loading) {
    return (
      <div className="security-monitor">
        <div className="loading">Carregando dados de segurança...</div>
      </div>
    );
  }

  return (
    <div className="security-monitor">
      <div className="security-header">
        <h2>🔒 Monitoramento de Segurança</h2>
        <button onClick={loadSecurityData} className="refresh-btn">
          🔄 Atualizar
        </button>
      </div>

      {/* Tabs */}
      <div className="security-tabs">
        <button
          className={activeTab === 'overview' ? 'tab active' : 'tab'}
          onClick={() => setActiveTab('overview')}
        >
          Visão Geral
        </button>
        <button
          className={activeTab === 'waf' ? 'tab active' : 'tab'}
          onClick={() => setActiveTab('waf')}
        >
          WAF
        </button>
        <button
          className={activeTab === 'integrity' ? 'tab active' : 'tab'}
          onClick={() => setActiveTab('integrity')}
        >
          Integridade
        </button>
        <button
          className={activeTab === 'vulnerabilities' ? 'tab active' : 'tab'}
          onClick={() => setActiveTab('vulnerabilities')}
        >
          Vulnerabilidades
        </button>
        <button
          className={activeTab === 'recommendations' ? 'tab active' : 'tab'}
          onClick={() => setActiveTab('recommendations')}
        >
          Recomendações
        </button>
      </div>

      {/* Overview Tab */}
      {activeTab === 'overview' && status && (
        <div className="tab-content">
          <div className="security-cards">
            {/* WAF Status */}
            <div className="security-card">
              <div className="card-header">
                <h3>{getStatusIcon(status.waf.status)} WAF (Firewall)</h3>
                <span className={`status-badge ${status.waf.status === 'active' ? 'active' : 'disabled'}`}>
                  {status.waf.status === 'active' ? 'ATIVO' : 'DESABILITADO'}
                </span>
              </div>
              <div className="card-body">
                <p><strong>Proteções Ativas:</strong></p>
                <ul>
                  {status.waf.protections.map((protection, idx) => (
                    <li key={idx}>✓ {protection}</li>
                  ))}
                </ul>
              </div>
            </div>

            {/* Integrity Status */}
            <div className="security-card">
              <div className="card-header">
                <h3>{getStatusIcon(status.integrity.status)} Integridade</h3>
                <span className={`status-badge ${status.integrity.status}`}>
                  {status.integrity.status.toUpperCase()}
                </span>
              </div>
              <div className="card-body">
                <p><strong>Arquivos Monitorados:</strong> {status.integrity.files_monitored}</p>
                {status.integrity.last_check && (
                  <p><strong>Última Verificação:</strong> {new Date(status.integrity.last_check).toLocaleString()}</p>
                )}
              </div>
            </div>

            {/* Vulnerabilities Status */}
            <div className="security-card">
              <div className="card-header">
                <h3>{getStatusIcon(status.vulnerabilities.status)} Vulnerabilidades</h3>
                <span className={`status-badge ${status.vulnerabilities.status}`}>
                  {status.vulnerabilities.issues_found} ENCONTRADAS
                </span>
              </div>
              <div className="card-body">
                {status.vulnerabilities.last_scan && (
                  <p><strong>Último Scan:</strong> {new Date(status.vulnerabilities.last_scan).toLocaleString()}</p>
                )}
                <p><strong>Issues:</strong> {status.vulnerabilities.issues_found}</p>
              </div>
            </div>

            {/* Compliance Status */}
            <div className="security-card">
              <div className="card-header">
                <h3>✅ Conformidade</h3>
                <span className="status-badge active">100%</span>
              </div>
              <div className="card-body">
                <p>✓ LGPD: {status.compliance.lgpd}</p>
                <p>✓ ISO 27001: {status.compliance.iso27001}</p>
                <p>✓ OWASP Top 10: {status.compliance.owasp}</p>
              </div>
            </div>
          </div>
        </div>
      )}

      {/* WAF Tab */}
      {activeTab === 'waf' && wafStats && (
        <div className="tab-content">
          <div className="waf-stats">
            <h3>Estatísticas do WAF</h3>
            <div className="stats-grid">
              <div className="stat-item">
                <div className="stat-value">{wafStats.requests_blocked}</div>
                <div className="stat-label">Requisições Bloqueadas</div>
              </div>
              <div className="stat-item">
                <div className="stat-value">{wafStats.sql_injection_attempts}</div>
                <div className="stat-label">Tentativas SQL Injection</div>
              </div>
              <div className="stat-item">
                <div className="stat-value">{wafStats.xss_attempts}</div>
                <div className="stat-label">Tentativas XSS</div>
              </div>
              <div className="stat-item">
                <div className="stat-value">{wafStats.rate_limit_violations}</div>
                <div className="stat-label">Violações Rate Limit</div>
              </div>
              <div className="stat-item">
                <div className="stat-value">{wafStats.blacklisted_ips}</div>
                <div className="stat-label">IPs Bloqueados</div>
              </div>
              <div className="stat-item">
                <div className="stat-value">{wafStats.total_requests}</div>
                <div className="stat-label">Total de Requisições</div>
              </div>
            </div>
          </div>
        </div>
      )}

      {/* Integrity Tab */}
      {activeTab === 'integrity' && integrityStatus && (
        <div className="tab-content">
          <div className="integrity-status">
            <h3>Status de Integridade</h3>
            {integrityStatus.status === 'not_configured' ? (
              <div className="alert alert-warning">
                <p>{integrityStatus.message}</p>
                <button className="btn-primary">Gerar Checksums</button>
              </div>
            ) : (
              <div className="integrity-info">
                <p><strong>Status:</strong> {integrityStatus.status}</p>
                <p><strong>Arquivos Monitorados:</strong> {integrityStatus.total_files}</p>
                <p><strong>Gerado em:</strong> {new Date(integrityStatus.generated_at).toLocaleString()}</p>
                <button className="btn-primary" onClick={() => alert('Verificação em desenvolvimento')}>
                  Verificar Integridade
                </button>
              </div>
            )}
          </div>
        </div>
      )}

      {/* Vulnerabilities Tab */}
      {activeTab === 'vulnerabilities' && vulnerabilities && (
        <div className="tab-content">
          <div className="vulnerabilities-status">
            <h3>Status de Vulnerabilidades</h3>
            {vulnerabilities.status === 'not_scanned' ? (
              <div className="alert alert-warning">
                <p>{vulnerabilities.message}</p>
                <button className="btn-primary">Executar Scan</button>
              </div>
            ) : (
              <div className="vuln-info">
                <p><strong>Último Scan:</strong> {new Date(vulnerabilities.scan_date).toLocaleString()}</p>
                
                <div className="vuln-section">
                  <h4>Python Dependencies</h4>
                  <p>Status: {vulnerabilities.python.status}</p>
                </div>
                
                <div className="vuln-section">
                  <h4>Node.js Dependencies</h4>
                  <p>Status: {vulnerabilities.nodejs.status}</p>
                </div>
                
                <div className="vuln-section">
                  <h4>Docker Images</h4>
                  <p>Status: {vulnerabilities.docker.status}</p>
                </div>
                
                <button className="btn-primary" onClick={() => alert('Scan em desenvolvimento')}>
                  Executar Novo Scan
                </button>
              </div>
            )}
          </div>
        </div>
      )}

      {/* Recommendations Tab */}
      {activeTab === 'recommendations' && (
        <div className="tab-content">
          <div className="recommendations">
            <h3>Recomendações de Segurança</h3>
            {recommendations.length === 0 ? (
              <div className="alert alert-success">
                ✅ Nenhuma recomendação pendente. Sistema seguro!
              </div>
            ) : (
              <div className="recommendations-list">
                {recommendations.map((rec, idx) => (
                  <div key={idx} className="recommendation-item" style={{ borderLeftColor: getPriorityColor(rec.priority) }}>
                    <div className="rec-header">
                      <h4>{rec.title}</h4>
                      <span className="priority-badge" style={{ backgroundColor: getPriorityColor(rec.priority) }}>
                        {rec.priority.toUpperCase()}
                      </span>
                    </div>
                    <p className="rec-description">{rec.description}</p>
                    <p className="rec-action"><strong>Ação:</strong> {rec.action}</p>
                  </div>
                ))}
              </div>
            )}
          </div>
        </div>
      )}
    </div>
  );
};

export default SecurityMonitor;
